org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/ClasspathJar.java

```diff
@@ -49,80 +51,60 @@
 import org.eclipse.jdt.internal.compiler.util.SuffixConstants;
 import org.eclipse.jdt.internal.core.JavaModelManager;
 import org.eclipse.jdt.internal.core.util.Util;
 
 public class ClasspathJar extends ClasspathLocation {
 final boolean isOnModulePath;
 
-static class PackageCacheEntry {
-	WeakReference<ZipFile> zipFile;
-	long lastModified;
-	long fileSize;
-	SimpleSet packageSet;
-
-	PackageCacheEntry(ZipFile zipFile, long lastModified, long fileSize, SimpleSet packageSet) {
-		this.zipFile = new WeakReference<>(zipFile);
-		this.lastModified = lastModified;
-		this.fileSize = fileSize;
-		this.packageSet = packageSet;
-	}
+private static record PackageCacheEntry(
+	WeakReference<ZipFile> zipFile,
+	long lastModified,
+	long fileSize,
+	Set<String> packageSet) {
 }
 
-protected static Map<String, PackageCacheEntry> PackageCache = new ConcurrentHashMap<>();
+protected static final Map<String, PackageCacheEntry> packageCache = new ConcurrentHashMap<>();
 
-protected static void addToPackageSet(SimpleSet packageSet, String fileName, boolean endsWithSep) {
+protected static void addToPackageSet(Set<String> packageSet, String fileName, boolean endsWithSep) {
 	int last = endsWithSep ? fileName.length() : fileName.lastIndexOf('/');
 	while (last > 0) {
 		// extract the package name
 		String packageName = fileName.substring(0, last);
-		if (packageSet.addIfNotIncluded(packageName) == null)
+		if (!packageSet.add(packageName)) {
 			return; // already existed
+		}
 		last = packageName.lastIndexOf('/');
 	}
 }
 
-/**
- * Calculate and cache the package list available in the zipFile.
- * @return A SimpleSet with the all the package names in the zipFile.
- */
-protected SimpleSet findPackageSet() {
-	PackageCacheEntry entry = PackageCache.compute(this.zipFilename, (zipFileName, cacheEntry) -> {
+private Set<String> getCachedPackageNames() {
+	PackageCacheEntry entry = packageCache.compute(this.zipFilename, (zipFileName, cacheEntry) -> {
 		if(cacheEntry != null && cacheEntry.zipFile.get() == this.zipFile) {
 			return cacheEntry;
 		}
 		long timestamp = this.lastModified();
 		if (cacheEntry != null && cacheEntry.lastModified == timestamp && cacheEntry.fileSize == this.fileSize) {
-			cacheEntry.zipFile = new WeakReference<>(this.zipFile);
-			return cacheEntry;
+			// cacheEntry.zipFile.get() != this.zipFile => update zipFile
+			return new PackageCacheEntry(new WeakReference<>(this.zipFile), cacheEntry.lastModified , cacheEntry.fileSize, cacheEntry.packageSet);
 		}
-		final SimpleSet packageSet = new SimpleSet(41);
-		packageSet.add(""); //$NON-NLS-1$
-		readJarContent(packageSet);
-		return new PackageCacheEntry(this.zipFile, timestamp, this.fileSize, packageSet);
+		return new PackageCacheEntry(new WeakReference<>(this.zipFile), timestamp, this.fileSize, Set.copyOf(readPackageNames()));
 	});
 
 	return entry.packageSet;
 }
-protected String readJarContent(final SimpleSet packageSet) {
-	String modInfo = null;
+/** overloaded */
```

Comment:
@jukzi  this comment looks plain wrong.

Additionally a reference to the issue / PR is missing from the commit comment.